### PR TITLE
optimize internal swap

### PR DESCRIPTION
**Pull request by wangjj9219**
1. cdp-treausury swap collateral to stable by the path with best price.
2. simplify code

### CODE REVIEW

**Review by xlc — APPROVED** [review]

**Comment by xlc** [comment]
Needs rebench

File: runtime/mandala/src/benchmarking/honzon.rs [hunk 1]
```diff
@@ -275,15 +212,10 @@ runtime_benchmarks! {
...
 
-		default_path.insert(0, currency_id);
-		for i in 0..default_path.len() {
-			if i != 0 {
-				inject_liquidity(maker.clone(), default_path[i], default_path[i-1], 10_000 * dollar(default_path[i]), 10_000 * dollar(default_path[i-1]), false)?;
-			}
-		}
+		// set balance and inject liquidity
+		set_balance(currency_id, &sender, (10 * collateral_amount) + ExistentialDeposits::get(&currency_id));
+		inject_liquidity(maker.clone(), currency_id, STABLECOIN, 10_000 * dollar(currency_id), 10_000 * dollar(STABLECOIN), false)?;
```
Comment (ferrell-code) [inline runtime/mandala/src/benchmarking/honzon.rs:218]: This makes it so the swap path is LIQUID => STABLECOIN in benchmark, when the worst case for acala/karura is LIQUID => STAKING => STABLECOIN. that will remove 2 read and 1 write from weights

File: runtime/mandala/src/benchmarking/cdp_engine.rs [hunk 2]
```diff
@@ -197,14 +195,8 @@ runtime_benchmarks! {
...
 
-		path.insert(0, LIQUID);
-		for i in 0..path.len() {
-			if i != 0 {
-				inject_liquidity(funder.clone(), path[i], path[i-1], 10_000 * dollar(path[i]), 10_000 * dollar(path[i-1]))?;
-			}
-		}
-
 		set_balance(LIQUID, &owner, (10 * collateral_amount) + ExistentialDeposits::get(&LIQUID));
+		inject_liquidity(funder, LIQUID, STABLECOIN, 10_000 * dollar(LIQUID), 10_000 * dollar(STABLECOIN))?;
```
Comment (ferrell-code) [inline runtime/mandala/src/benchmarking/cdp_engine.rs:200]: same here acala/karura worst case LIQUID => STAKING => STABLECOIN. We could always do a followup pr as it doesn't effect any  logic

File: modules/dex/src/lib.rs [hunk 3]
```diff
@@ -1227,34 +1227,87 @@ impl<T: Config> DEXManager<T::AccountId, CurrencyId, Balance> for Pallet<T> {
...
+		let default_swap_path = vec![supply_currency_id, target_currency_id];
+		let mut maybe_best = Self::get_swap_amount(&default_swap_path, limit)
+			.map(|(supply_amout, target_amount)| (default_swap_path, supply_amout, target_amount));
+
+		for path_joint in alternative_path_joint_list {
+			if !path_joint.is_empty() {
+				let mut swap_path = vec![];
+
+				if supply_currency_id != path_joint[0] {
+					swap_path.extend(vec![supply_currency_id]);
```
Comment (ferrell-code) [inline modules/dex/src/lib.rs:1272]:
```suggestion
					swap_path.push(supply_currency_id);
```

File: modules/dex/src/lib.rs [hunk 4]
```diff
@@ -1227,34 +1227,87 @@ impl<T: Config> DEXManager<T::AccountId, CurrencyId, Balance> for Pallet<T> {
...
-		who: &T::AccountId,
-		path: &[CurrencyId],
-		supply_amount: Balance,
-		min_target_amount: Balance,
-	) -> sp_std::result::Result<Balance, DispatchError> {
-		Self::do_swap_with_exact_supply(who, path, supply_amount, min_target_amount)
+				swap_path.extend(path_joint.clone());
+
+				if target_currency_id != path_joint[path_joint.len() - 1] {
+					swap_path.extend(vec![target_currency_id]);
```
Comment (ferrell-code) [inline modules/dex/src/lib.rs:1278]:
```suggestion
					swap_path.push(target_currency_id);
```

**Review by ferrell-code — APPROVED** [review]
lgtm, nits above

File: modules/dex/src/lib.rs (hunk 3 above)
Comment (ferrell-code) [reply modules/dex/src/lib.rs:1272]: Just syntax so doesn't matter to me if this is changed

**Review by xlc — APPROVED** [review]

**Comment by wangjj9219** [comment]
@qwer951123  `honzon.closeLoanHasDebitByDex` changed, plz update front end and sdk

**Pull request merged**